src/main/scala/com/fulcrumgenomics/bam/AddReadGroupsByName.scala

```diff
@@ -0,0 +1,182 @@
+ @arg(doc="The library to insert into the read group header") val library: String,
+ @arg(doc="The platform type (e.g. illumina, solid) to insert into the read group header") val platform: Option[String] = Some("ILLUMINA"),
+ @arg(doc="The sequencing center from which the data originated") val sequencingCenter: Option[String] = None,
+ @arg(doc="Predicted median insert size, to insert into the read group header") val predictedInsertSize: Option[Integer] = None,
+ @arg(doc="Program group to insert into the read group header.") val programGroup: Option[String] = None,
+ @arg(doc="Platform model to insert into the group header (free-form text providing further details of the platform/technology used)") val platformModel: Option[String] = None,
+ @arg(doc="Description inserted into the read group header") val description: Option[String] = None,
+ @arg(doc="Date the run was produced, to insert into the read group header") val runDate: Option[Iso8601Date] = None,
+ @arg(doc="Comment(s) to include in the merged output file's header.", minElements = 0) val comments: List[String] = Nil,
+ @arg(doc="The sort order for the output sam/bam file.") val sortOrder: SortOrder = SortOrder.queryname
+) extends FgBioTool with LazyLogging {
+
+  import scala.collection.JavaConversions.{asScalaIterator, seqAsJavaList}
+
+  Io.assertReadable(input)
+  Io.assertCanWriteFile(output)
+
+  private val nextId = new AtomicInteger(0)
+  private val readGroups = new mutable.HashMap[RunInfo, SAMReadGroupRecord]()
+
+  override def execute(): Unit = {
+
+    // Gather all the read groups
+    {
+      val progress = new ProgressLogger(logger, verb = "read", unit = 5e6.toInt)
+      val in = SamReaderFactory.make().open(input.toFile)
+
+      in.iterator().foreach { record =>
+        val runInfo       = RunInfo(name=record.getReadName)
+        val platformModel = this.platformModel.getOrElse(runInfo.platformModel)
+
+        if (!readGroups.contains(runInfo)) {
+          val readGroup = new SAMReadGroupRecord(nextId.incrementAndGet().toString)
+
+          readGroup.setSample(sample)
+          readGroup.setLibrary(library)
+          readGroup.setPlatformUnit(runInfo.platformUnit)
+          readGroup.setPlatformModel(platformModel)
+          platform.foreach(readGroup.setPlatform)
+          sequencingCenter.foreach(readGroup.setSequencingCenter)
+          predictedInsertSize.foreach(readGroup.setPredictedMedianInsertSize)
+          description.foreach(readGroup.setDescription)
+          runDate.foreach(readGroup.setRunDate)
+
+          readGroups(runInfo) = readGroup
+        }
+        else {
+          // Check that the read information matches the read group
+          val readGroup = readGroups(runInfo)
+
+          // Used in the lookup of the read group, should never happen!
+          if (readGroup.getPlatformUnit != runInfo.platformUnit) unreachable(s"PU mismatch: s'${readGroup.getPlatformModel} != ${runInfo.platformUnit}")
```

Comment:
See the comment on 113 ("should never happen").  I am removing this else branch as it should never happen, so it's too defensive and overkill.